File: users/phques/tools/convert_adaptives.cs
// Convert JSON adaptive trigger pairs into QMK adaptive_3key_t/adaptive_2key_t tables.
// Usage: cat rules.json | dotnet run --project convert_adaptives.csproj
// Or compiled: cat rules.json | ./convert_adaptives

using System.Text.Json;

var special = new Dictionary<char, string>
{
    ['.']  = "KC_DOT",
    ['/']  = "KC_SLASH",
    ['"']  = "KC_DQUO",
    ['\''] = "KC_QUOTE",
    ['#']  = "KC_HASH",
};

string KeycodeForChar(char ch)
{
    char lower = char.ToLower(ch);
    if (lower >= 'a' && lower <= 'z')
        return "KC_" + char.ToUpper(lower);
    if (special.TryGetValue(lower, out var kc))
        return kc;
    throw new ArgumentException($"Unsupported key char: '{ch}'");
}

string CEscape(string s) =>
    s.Replace("\\", "\\\\").Replace("\"", "\\\"");

string MappingComment(string trigger, string output) =>
    $"// {trigger} -> {output}";

string FormatEntry(string entry, string comment, int width)
{
    var withComma = (entry + ",").PadRight(width + 1);
    return $"    {withComma} {comment}";
}

List<(string trigger, string output)> ParseRules(string input)
{
    var data = JsonSerializer.Deserialize<List<List<string>>>(input)
        ?? throw new InvalidOperationException("Top-level JSON must be an array");

    var parsed = new List<(string, string)>();
    foreach (var item in data)
    {
        if (item.Count != 2)
            throw new ArgumentException($"Each row must be [trigger, output], got: [{string.Join(", ", item)}]");

        var trig = item[0];
        var output = item[1];

        if (trig.Length != 2 && trig.Length != 3)
            throw new ArgumentException($"Trigger must be length 2 or 3, got: \"{trig}\"");

        parsed.Add((trig, output));
    }
    return parsed;
}

void EmitAdaptiveTables(List<(string trigger, string output)> rules)
{
    var threeKey = rules.Where(r => r.trigger.Length == 3).ToList();
    var twoKey   = rules.Where(r => r.trigger.Length == 2).ToList();

    var threePrefixes = new HashSet<(char, char)>(
        threeKey.Select(r => (r.trigger[0], r.trigger[1])));

    var threeEntries = threeKey
        .Select(r => (
            entry:   $"{{{KeycodeForChar(r.trigger[0])}, {KeycodeForChar(r.trigger[1])}, {KeycodeForChar(r.trigger[2])}, \"{CEscape(r.output)}\"}}",
            comment: MappingComment(r.trigger, r.output)))
        .ToList();

    var twoEntries = twoKey
        .Select(r => (
            entry:   $"{{{KeycodeForChar(r.trigger[0])}, {KeycodeForChar(r.trigger[1])}, \"{CEscape(r.output)}\"}}",
            comment: MappingComment(r.trigger, r.output)))
        .ToList();

    int threeWidth = threeEntries.Count > 0 ? threeEntries.Max(e => e.entry.Length) : 0;
    int twoWidth   = twoEntries.Count   > 0 ? twoEntries.Max(e => e.entry.Length)   : 0;

    Console.WriteLine("// Auto-generated adaptive key tables. Do not edit directly.");
    Console.WriteLine("#include <quantum.h>");
    Console.WriteLine();
    Console.WriteLine("#include \"../../processAdaptive.h\"");
    Console.WriteLine();
    Console.WriteLine("const adaptive_3key_t adaptive_3keys[] PROGMEM = {");
    foreach (var (entry, comment) in threeEntries)
        Console.WriteLine(FormatEntry(entry, comment, threeWidth));
    Console.WriteLine("    {0, 0, 0, NULL}");
    Console.WriteLine("};");
    Console.WriteLine();

    Console.WriteLine("const adaptive_2key_t adaptive_2keys[] PROGMEM = {");
    foreach (var (entry, comment) in twoEntries)
        Console.WriteLine(FormatEntry(entry, comment, twoWidth));
    Console.WriteLine("    {0, 0, NULL}");
    Console.WriteLine("};");

    var conflicts = twoKey
        .Where(r => threePrefixes.Contains((r.trigger[0], r.trigger[1])))
        .OrderBy(r => r.trigger)
        .ToList();

    if (conflicts.Count > 0)
    {
        Console.Error.WriteLine();
        Console.Error.WriteLine("WARNING: Shared-prefix conflicts found:");
        foreach (var r in conflicts)
            Console.Error.WriteLine($"  2-key trigger \"{r.trigger}\" -> \"{r.output}\" shares prefix with a 3-key trigger.");
    }
}

try
{
    var input = Console.In.ReadToEnd();
    var rules = ParseRules(input);
    EmitAdaptiveTables(rules);
}
catch (Exception ex)
{
    Console.Error.WriteLine($"Error: {ex.Message}");
    Environment.Exit(1);
}

File: users/phques/tools/convert_adaptives.py
#!/usr/bin/env python3
"""Convert JSON adaptive trigger pairs into QMK adaptive_3key_t/adaptive_2key_t tables.

Input format (stdin):
[
  ["mwm", "mpl"],
  ["wm", "lm"]
]

Output format:
const adaptive_3key_t adaptive_3keys[] = {
    {KC_M, KC_W, KC_M, "mpl"},
    {0, 0, 0, NULL}
};

const adaptive_2key_t adaptive_2keys[] = {
    {KC_W, KC_M, "lm"},
    {0, 0, NULL}
};

Rules:
- 3-char triggers are emitted into adaptive_3keys.
- 2-char triggers are emitted into adaptive_2keys.
- Warn if any 2-char trigger shares a prefix with a 3-char trigger.
"""

import json
import string
import sys
from typing import List, Tuple

SPECIAL = {
    ".": "KC_DOT",
    ",": "KC_COMMA",
    "/": "KC_SLASH",
    "=": "KC_EQUAL",
    '"': "KC_DQUO",
    "'": "KC_QUOTE",
    "#": "KC_HASH",
    "-": "KC_MINUS",
}

LETTERS = {c: f"KC_{c.upper()}" for c in string.ascii_lowercase}


def keycode_for_char(ch: str) -> str:
    c = ch.lower()
    if c in LETTERS:
        return LETTERS[c]
    if c in SPECIAL:
        return SPECIAL[c]
    raise ValueError(f"Unsupported key char: {ch!r}")


def c_escape(s: str) -> str:
    return s.replace("\\", "\\\\").replace('"', '\\"')


def mapping_comment(trigger: str, output: str) -> str:
    # Emit block comments to avoid C line-splice issues from trailing backslashes.
    safe_output = output.replace("\n", "\\n").replace("\r", "\\r").replace("*/", "* /")
    return f"/* {trigger} -> {safe_output} */"


def format_entry(entry: str, comment: str, width: int) -> str:
    return f"    {(entry + ',').ljust(width + 1)} {comment}"


def parse_rules(data) -> List[Tuple[str, str]]:
    if not isinstance(data, list):
        raise ValueError("Top-level JSON must be an array")

    parsed: List[Tuple[str, str]] = []
    for item in data:
        if not (isinstance(item, list) and len(item) == 2):
            raise ValueError(f"Each row must be [trigger, output], got: {item!r}")

        trig, out = item
        if not (isinstance(trig, str) and isinstance(out, str)):
            raise ValueError(f"Trigger and output must be strings, got: {item!r}")

        if len(trig) not in (2, 3):
            raise ValueError(f"Trigger must be length 2 or 3, got: {trig!r}")

        parsed.append((trig, out))

    return parsed


def emit_adaptive_tables(rules: List[Tuple[str, str]]) -> None:
    three_key = [(trig, out) for trig, out in rules if len(trig) == 3]
    two_key = [(trig, out) for trig, out in rules if len(trig) == 2]

    three_prefixes = {(trig[0], trig[1]) for trig, _ in three_key}
    three_entries = [
        (
            f'{{{keycode_for_char(trig[0])}, {keycode_for_char(trig[1])}, {keycode_for_char(trig[2])}, "{c_escape(out)}"}}',
            mapping_comment(trig, out),
        )
        for trig, out in three_key
    ]
    two_entries = [
        (
            f'{{{keycode_for_char(trig[0])}, {keycode_for_char(trig[1])}, "{c_escape(out)}"}}',
            mapping_comment(trig, out),
        )
        for trig, out in two_key
    ]
    three_width = max((len(entry) for entry, _ in three_entries), default=0)
    two_width = max((len(entry) for entry, _ in two_entries), default=0)

    print("// Auto-generated adaptive key tables. Do not edit directly.")
    print("#include <quantum.h>\n")
    print("#include \"../../processAdaptive.h\"\n")
    print("const adaptive_3key_t adaptive_3keys[] PROGMEM = {")

    for entry, comment in three_entries:
        print(format_entry(entry, comment, three_width))

    print("    {0, 0, 0, NULL}")
    print("};")
    print("")

    print("const adaptive_2key_t adaptive_2keys[] PROGMEM = {")
    for entry, comment in two_entries:
        print(format_entry(entry, comment, two_width))

    print("    {0, 0, NULL}")
    print("};")

    conflicts = [(trig, out) for trig, out in two_key if (trig[0], trig[1]) in three_prefixes]
    if conflicts:
        print("", file=sys.stderr)
        print("WARNING: Shared-prefix conflicts found:", file=sys.stderr)
        for trig, out in conflicts:
            print(
                f"  2-key trigger {trig!r} -> {out!r} shares prefix with a 3-key trigger.",
                file=sys.stderr,
            )


def main() -> int:
    try:
        data = json.load(sys.stdin)
        rules = parse_rules(data)
        emit_adaptive_tables(rules)
    except Exception as exc:  # keep CLI output simple for shell usage
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
